apollo-router/src/spec/query/change.rs

```diff
@@ -348,16 +501,41 @@ impl<'a> Hasher for QueryHashVisitor<'a> {
     }
 
     fn write(&mut self, bytes: &[u8]) {
+        // byte separator between each part that is hashed
+        self.hasher.update(&[0xFF][..]);
         self.hasher.update(bytes);
     }
 }
 
 impl<'a> Visitor for QueryHashVisitor<'a> {
     fn operation(&mut self, root_type: &str, node: &executable::Operation) -> Result<(), BoxError> {
+        "^VISIT_OPERATION".hash(self);
+
         root_type.hash(self);
         self.hash_type_by_name(root_type)?;
+        node.operation_type.hash(self);
+        node.name.hash(self);
+        for variable in &node.variables {
+            variable.name.hash(self);
+            self.hash_type(&variable.ty)?;
+
+            if let Some(value) = variable.default_value.as_ref() {
+                self.hash_value(value);
+            } else {
+                "^VISIT_OPERATION-NO_DEFAULT".hash(self);
+            }
+
+            for directive in &variable.directives {
+                self.hash_directive(directive);
+            }
+        }
+        for directive in &node.directives {
+            self.hash_directive(directive);
+        }
```

Comment:
It's vulnerable to collisions, I wrote test for it:
```rust
    #[test]
    fn it_change_directive_location() {
        let schema: &str = r#"
        directive @foo on QUERY | VARIABLE_DEFINITION

        type Query {
          field(arg: String): String
        }
        "#;

        let query_one = r#"
        query Test ($arg: String @foo) {
          field(arg: $arg) 
        }
        "#;
        let query_two = r#"
        query Test ($arg: String) @foo {
          field(arg: $arg) 
        }
        "#;

        assert_ne!(
            hash(schema, query_one).from_hash_query,
            hash(schema, query_two).from_hash_query
        );
        assert_ne!(
            hash(schema, query_one).from_visitor,
            hash(schema, query_two).from_visitor
        );
    }

```

Reply:
Some places have similar code where stuff is hashed in a cycle, but they are not prefixed by anything as a group.
I can't trigger collision in those places because they are not followed by stuff that can cause collision but that could change in future.
For example:
https://github.com/apollographql/router/blob/81604c235b8c7e601204e804194bb381df576d71/apollo-router/src/spec/query/change.rs#L319-L324
Currently you can't cause a collision, but it's better to still wrap it in a "input field definition" tag.

Reply:
so idea is if you have has of multiple stuff in a cycle those should be prefixed as a group.

Reply:
that's a good point. This should be fixed in [1b76c0a](https://github.com/apollographql/router/pull/5255/commits/1b76c0ad01743272dfaf4497d87760ea9d265b1f)